dffml/util/testing/manifest/shim.py

```diff
@@ -0,0 +1,704 @@
+forward.
+
+To make sure we have forwards / backwards compatibility we should
+include information which allows us to identify what format the document
+is in, and what version of that format it is. This will likely also feed
+into our input dataflow requirements as we'll need to have the ability
+to check an arbitrary input to see if we might have an applicable
+converter.
+
+Let's learn from JSON Schema and include a URL where we might be able
+to find the schema for the document. We can double up on our previous
+needs by asking that the filename of the URL can help us identify our
+document format (we'll provide fallback for if we don't have control
+over the filename via the ``document_format`` and ``$document_version``
+keys). We'll parse the URL for the filename component. When we parse it
+we'll split on ``.``. If the first part is eff (Extensible Format
+Format) we'll treat the rest up until the semantic version as the format
+name. Then the semantic version is the version of the format. Then the
+rest should be the extension which is associated with the format which
+we can use to validate the contents of the document, such as JSON
+schema.
+
+``$schema: "https://example.com/eff.my.document.format.0.0.0.schema.json"``
+
+TODO
+----
+
+- Verification of the manifest. Idea: Developer generates manifest.
+  Signs manifest with public asymmetric key. Prepends base64 encoded
+  signature as a valid key, ``$signature``. This means you have to
+  parse the YAML before you have verified the signature, which is not
+  ideal. However, it's one method available to us and a simple parse
+  without the use of a full YAML parser could be done. Or we could
+  distribute out of band and verify the document before the conversion
+  stage, in the loading stage.
+
+- Verification of references within manifest. Do we support public
+  portion of CA key embedded in the document various places? We
+  could then use it for things like verification of git repos where
+  the CA must sign all developer keys which are in the repo history.
+  This will apply to anything that is an external reference in the
+  document. There should be a way for the document to include an HMAC or
+  something like that or something more dynamic like a CA.
+
+Notes
+-----
+
+- SSH public keys: https://github.com/$USERNAME.keys
+
+- https://github.com/mjg59/ssh_pki
+"""
+import os
+import sys
+import hmac
+import json
+import shlex
+import pickle
+import pathlib
+import hashlib
+import argparse
+import functools
+import importlib
+import traceback
+import contextlib
+import subprocess
+import dataclasses
+from typing import Dict, List, Callable, Any, Union, Optional
+
+
+def popen_write_to_stdin(
+    cmd: List[str], write: bytes, **kwargs
+) -> subprocess.Popen:
+    """
+    Call :py:func:`subprocess.Popen`
+    """
+    read_end, write_end = os.pipe()
+    proc = subprocess.Popen(cmd, stdin=read_end, **kwargs)
```

Comment:
doctest and `Popen.communicate` were not playing nice together, so when with pipe for now